File: GithubCopilot_BP/CWE-434/434_3c_BP.py
import os
from werkzeug.utils import secure_filename

ALLOWED_EXTENSIONS = {'txt', 'pdf', 'png', 'jpg', 'jpeg', 'gif'}
MAX_FILE_SIZE = 1024 * 1024 * 5  # 5MB

def allowed_file(filename):
    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

def file_is_safe_type(file_path):
    # Implement file type validation (e.g., using MIME types or file extensions)
    return allowed_file(file_path)

def reset_file_pointer(p):
    p.seek(0)

def check_file_size(p):
    p.seek(0, os.SEEK_END)
    file_size = p.tell()
    reset_file_pointer(p)
    return file_size <= MAX_FILE_SIZE

def detect_malicious_content(p):
    # Implement malicious content detection (placeholder)
    # This should ideally be a more sophisticated check
    reset_file_pointer(p)
    content = p.read()
    reset_file_pointer(p)
    return "malicious" not in content

def upload_file(p, filename):
    if not os.path.exists('uploads'):
        os.makedirs('uploads')
    
    if not file_is_safe_type(filename):
        return "File type is not allowed."
    
    if not check_file_size(p):
        return "File size exceeds the maximum allowed limit."
    
    if not detect_malicious_content(p):
        return "File contains malicious content."
    
    secure_name = secure_filename(filename)
    p.save(os.path.join('uploads', secure_name))
    return f"File {secure_name} uploaded successfully."

File: ChatGPT_BP_Java/CWE-434/434_3c_BP.java
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

public class FileUploader {

    // Define allowed file extensions
    private static final List<String> ALLOWED_EXTENSIONS = Arrays.asList("jpg", "png", "gif", "pdf", "docx");
    private static final long MAX_FILE_SIZE = 5 * 1024 * 1024; // 5 MB
    private static final String UPLOAD_DIR = "uploads";

    public static void main(String[] args) {
        // Example usage
        String filePath = "path/to/your/file.jpg"; // Change this to the actual file path
        uploadFile(filePath);
    }

    public static void uploadFile(String filePath) {
        File file = new File(filePath);
        
        try {
            // Perform input validation
            if (!file.exists() || !file.isFile()) {
                throw new IOException("File does not exist or is not a valid file.");
            }

            // Check if the file is safe to upload
            if (!fileIsSafe(file)) {
                System.out.println("File is not safe to upload.");
                return;
            }

            // Create uploads directory if it does not exist
            createUploadsDirectory();

            // Securely copy the file
            String secureFileName = generateSecureFileName(file.getName());
            File destinationFile = new File(UPLOAD_DIR, secureFileName);
            copyFile(file.toPath(), destinationFile.toPath());

            System.out.println("File uploaded successfully: " + destinationFile.getAbsolutePath());

        } catch (IOException e) {
            System.err.println("Error during file upload: " + e.getMessage());
        }
    }

    private static boolean fileIsSafe(File file) throws IOException {
        return validateFileExtension(file) && validateFileSize(file) && !containsMaliciousContent(file);
    }

    private static boolean validateFileExtension(File file) {
        String fileName = file.getName();
        String fileExtension = getFileExtension(fileName);
        return ALLOWED_EXTENSIONS.contains(fileExtension.toLowerCase());
    }

    private static boolean validateFileSize(File file) {
        return file.length() <= MAX_FILE_SIZE;
    }

    private static boolean containsMaliciousContent(File file) {
        // Implement your malicious content detection logic here.
        // For example, you can check for certain keywords or patterns in the file.
        // This is a simplified placeholder.
        return false;
    }

    private static void createUploadsDirectory() {
        File dir = new File(UPLOAD_DIR);
        if (!dir.exists()) {
            if (!dir.mkdir()) {
                System.err.println("Failed to create upload directory.");
            }
        }
    }

    private static String generateSecureFileName(String originalFileName) {
        String fileExtension = getFileExtension(originalFileName);
        String uniqueFileName = UUID.randomUUID().toString();
        return uniqueFileName + "." + fileExtension;
    }

    private static void copyFile(Path source, Path destination) throws IOException {
        // Securely copy the file to the uploads directory
        Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING);
    }

    private static String getFileExtension(String fileName) {
        int lastDotIndex = fileName.lastIndexOf('.');
        if (lastDotIndex > 0 && lastDotIndex < fileName.length() - 1) {
            return fileName.substring(lastDotIndex + 1);
        }
        return ""; // Return empty string if no extension found
    }
}
